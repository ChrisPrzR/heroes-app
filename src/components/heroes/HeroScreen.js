import React from 'react';
import { useParams, Redirect } from 'react-router-dom';


import { getHeroeById } from '../../selectors/getHeroById'

export const HeroScreen = ({history}) => {

    const {heroeId} = useParams();

    const hero = getHeroeById( heroeId);

    if(!hero){
        return <Redirect to="/" />;
    }

    const handleReturn = () => {
        if( history.length <=2 ){
            hero.publisher === 'Marvel Comics' && history.push('/');
            hero.publisher === 'DC Comics' && history.push('/dc');
        }else{
        history.goBack();
        }
    }

    const {
        superhero,
        alter_ego,
        first_appearance,
        characters} = hero;

    return (
        <div className="card ms-3" style={ {maxWidth: '540px', maxHeight: '80vh' } }>
            <div className="row no-gutters">
                <div className="col-md-4" style={{height: '100%'}}>
                    <img src={ `../assets/heroes/${ heroeId }.jpg` } style={{height: '100%'}} className="card-img" alt={superhero}/>
                </div>
                <div className="col-md-8">
                    <div className="card-body">
                        <h5 className="card-title">{superhero}</h5>
                        <p className="card-text">{ alter_ego }</p>

                        {
                            (alter_ego !== characters)
                            && <p className="card-text">{characters}</p>
                        }

                        <p className="card-text">
                            <small className="text-muted"> { first_appearance }</small>
                        </p>

                        <h5>Characters</h5>
                        <p>{characters}</p>
                        <button 
                            className="btn btn-outline-info"
                            onClick={handleReturn}>
                            Return
                        </button>
                    </div>
                </div>
                
            </div>
        </div>
    );
}
