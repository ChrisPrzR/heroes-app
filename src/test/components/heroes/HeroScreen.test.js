import React from 'react';
import { mount } from 'enzyme'
import { HeroScreen } from '../../../components/heroes/HeroScreen';
import { MemoryRouter, Route } from 'react-router-dom';


describe('Testing <HeroScreen/>', () => {
    
    const history = {
        length: 10,
        push: jest.fn(),
        goBack: jest.fn()
    }

    
    test('should render redirect component if not arguments in URL exist', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero']}>
                <HeroScreen history={history} />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot()
    })

    test('should render a hero if param exists and valid', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-thor']}>
                <Route path="/hero/:heroeId" component={HeroScreen}/>                    
            </MemoryRouter>
        );

        expect(wrapper.find('.row').exists()).toBe(true)
    })
    
    test('should return to previous screen PUSH', () => {
        
        const history = {
            length: 1,
            push: jest.fn(),
            goBack: jest.fn()
        }

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-thor']}>
                <Route 
                    path="/hero/:heroeId" 
                    component={(props) => <HeroScreen history={history} />}
                />                    
            </MemoryRouter>
        );

        wrapper.find('button').prop('onClick')();

        expect(history.push).toHaveBeenCalledWith('/')
        expect(history.goBack).not.toHaveBeenCalled()
    })

    test('should return to previous screen GOBACK', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-thor']}>
                <Route
                    path="/hero/:heroeId"
                    component={() => <HeroScreen history={history}/>}
                />
            </MemoryRouter>

        );

        wrapper.find('button').simulate('click');
        
        expect(history.push).not.toHaveBeenCalled()
        expect(history.goBack).toHaveBeenCalled()
    })
    
    
    
})
