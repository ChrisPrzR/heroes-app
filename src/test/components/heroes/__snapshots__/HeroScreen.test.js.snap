// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <HeroScreen/> should render redirect component if not arguments in URL exist 1`] = `""`;
